### PR TITLE
Add a typeof check to Object.getPrototypeOf. [closes #319, #320]

### LINKED ISSUE #319: Object.getPrototypeOf constructor check.
I noticed the current `Object.getPrototypeOf` check:
```js
if (object.constructor)
```
but the constructor can really be any value so maybe a more thorough check like
```js
if (typeof object.constructor === 'function')
```
would be better.

### PR DESCRIPTION
**Pull request by jdalton**
See #319

### CODE REVIEW

**Comment by jdalton** [comment]
Travis test fails due to npm.

File: es5-sham.js [hunk 1]
```diff
@@ -65,7 +65,7 @@ if (!Object.getPrototypeOf) {
         /* eslint-enable no-proto */
         if (proto || proto === null) {
             return proto;
-        } else if (object.constructor) {
+        } else if (typeof object.constructor === 'function') {
```
Comment (ljharb) [inline es5-sham.js]: Will this not return a false positive with regexes in some older engines?

I think we need an `Object#toString` check here instead.

**Comment by ljharb** [comment]
Can we add a test with `constructor` set to a non-function, to prevent a regression?

**Comment by jdalton** [comment]
Ah nice!

**Comment by jdalton** [comment]
There are currently no tests for `Object.getPrototypeOf`.

File: es5-sham.js (hunk 1 above)
Comment (jdalton) [inline es5-sham.js]: Correct. I saw `typeof` checks in other places so wasn't sure your chance on it.

**Comment by jdalton** [comment]
Is there a way to detect if the test is running in enviros where the built-in exists and isn't shimmed?

**Comment by ljharb** [comment]
Any tests should run only when the builtin exists, and shouldn't care whether it's shimmed or not, ideally.

**Comment by jdalton** [comment]
Yap, gotta test the sham behavior which means different behavior with and without the native built-in.

File: tests/spec/s-object.js [hunk 2]
```diff
@@ -260,6 +260,38 @@ describe('Object', function () {
...
+        });
+
+        it('should shamone to the `Object.prototype` if `object.constructor` is not a function', function () {
+            function Foo() {}
+            Foo.prototype.constructor = 1;
+
+            var proto = Object.getPrototypeOf(new Foo),
+                fnToString = Function.prototype.toString;
+
+            if (fnToString.call(Object.getPrototypeOf).indexOf('[native code]') < 0) {
```
Comment (ljharb) [inline tests/spec/s-object.js]: While this works, it makes me cringe - why is it that native code will return `Object.prototype` but the sham will return `Foo.prototype`? I'd rather the sham threw, than return the wrong value.

File: tests/spec/s-object.js [hunk 3]
```diff
@@ -260,6 +260,38 @@ describe('Object', function () {
         });
     });
 
+    describe('Object.getPrototypeOf', function () {
+        it('should return the [[Prototype]] of an object', function () {
+            function Foo() {}
```
Comment (ljharb) [inline tests/spec/s-object.js]: linting rules forbid function declarations - `var Foo = function Foo() {};` would work here.

File: tests/spec/s-object.js [hunk 4]
```diff
@@ -260,6 +260,38 @@ describe('Object', function () {
...
+
+            var proto = Object.getPrototypeOf(new Foo);
```
Comment (ljharb) [inline tests/spec/s-object.js]: `new Foo` needs to be `new Foo()`

File: tests/spec/s-object.js [hunk 5]
```diff
@@ -260,6 +260,38 @@ describe('Object', function () {
...
+            if (fnToString.call(Object.getPrototypeOf).indexOf('[native code]') < 0) {
+                expect(proto).toBe(Object.prototype);
+            } else {
+                expect(proto).toBe(Foo.prototype);
+            }
+        });
+
+        it('should throw error for non object', function () {
+            try {
+                expect(Object.getPropertyOf(1)).toBeUndefined();
```
Comment (ljharb) [inline tests/spec/s-object.js]: s/`getPropertyOf`/`getPrototypeOf`, otherwise you'll get "undefined is not a function" :-)

**Comment by ljharb** [comment]
I haven't looked at the sham code in awhile - is the difference in behavior because of unshimmable semantics?

**Comment by ljharb** [comment]
Would you classify this as a breaking change in the sham? or as a bug fix?

If it'll be a patch, let's rebase this and I'll merge and release it. If it's a breaking change, I'll be a bit more cautious.

**Comment by jdalton** [comment]
It's more a bug fix because before the constructor could have been a primitive which isn't the intent of the path.

**Pull request merged**